src/jit/liveness.cpp

```diff
@@ -546,65 +560,79 @@ void Compiler::fgPerBlockLocalVarLiveness()
-                }
+                    noway_assert(tree->gtOp.gtOp1);
+                    GenTreePtr rhsNode;
+                    if (tree->gtOper == GT_ASG)
+                    {
+                        noway_assert(tree->gtOp.gtOp2);
+                        asgdLclVar = tree->gtOp.gtOp1;
+                        rhsNode    = tree->gtOp.gtOp2;
+                    }
+                    else
+                    {
+                        asgdLclVar = tree;
+                        rhsNode    = tree->gtOp.gtOp1;
+                    }
 
-                // If this is an assignment to local var with no SIDE EFFECTS,
-                // set asgdLclVar so that genMarkUseDef will flag potential
-                // x=f(x) expressions as GTF_VAR_USEDEF.
-                // Reset the flag before recomputing it - it may have been set before,
-                // but subsequent optimizations could have removed the rhs reference.
-                asgdLclVar->gtFlags &= ~GTF_VAR_USEDEF;
-                if ((rhsNode->gtFlags & GTF_SIDE_EFFECT) == 0)
-                {
-                    noway_assert(asgdLclVar->gtFlags & GTF_VAR_DEF);
-                }
-                else
-                {
-                    asgdLclVar = nullptr;
+                    // If this is an assignment to local var with no SIDE EFFECTS,
+                    // set asgdLclVar so that genMarkUseDef will flag potential
+                    // x=f(x) expressions as GTF_VAR_USEDEF.
+                    // Reset the flag before recomputing it - it may have been set before,
+                    // but subsequent optimizations could have removed the rhs reference.
+                    asgdLclVar->gtFlags &= ~GTF_VAR_USEDEF;
+                    if ((rhsNode->gtFlags & GTF_SIDE_EFFECT) == 0)
+                    {
+                        noway_assert(asgdLclVar->gtFlags & GTF_VAR_DEF);
+                    }
+                    else
+                    {
+                        asgdLclVar = nullptr;
+                    }
                 }
-            }
 
 #ifdef LEGACY_BACKEND
-            tree = fgLegacyPerStatementLocalVarLiveness(stmt->gtStmt.gtStmtList, NULL, asgdLclVar);
+                tree = fgLegacyPerStatementLocalVarLiveness(stmt->gtStmt.gtStmtList, NULL, asgdLclVar);
 
-            // We must have walked to the end of this statement.
-            noway_assert(!tree);
-#else
-            fgPerStatementLocalVarLiveness(stmt->gtStmt.gtStmtList, asgdLclVar);
-#endif // LEGACY_BACKEND
+                // We must have walked to the end of this statement.
+                noway_assert(!tree);
+#else  // !LEGACY_BACKEND
+                fgPerStatementLocalVarLiveness(stmt->gtStmt.gtStmtList, asgdLclVar);
+#endif // !LEGACY_BACKEND
+            }
+        }
+        else
+        {
+#ifdef LEGACY_BACKEND
+            unreached();
+#else  // !LEGACY_BACKEND
+            // NOTE: the `asgdLclVar` analysis done above is not correct for LIR: it depends
```

Comment:
It may or may not be worth mentioning here that the `asgdlclVar` analysis is not needed in LIR because the nodes are traversed in an order that reflects the data flow.